lib/ocpp/v201/smart_charging.cpp

```diff
@@ -319,25 +329,39 @@ SmartChargingHandler::validate_profile_schedules(ChargingProfile& profile,
 SetChargingProfileResponse SmartChargingHandler::add_profile(int32_t evse_id, ChargingProfile& profile) {
     SetChargingProfileResponse response;
     response.status = ChargingProfileStatusEnum::Accepted;
+    auto found_profile = false;
+    for (auto& [existing_evse_id, evse_profiles] : charging_profiles) {
+        for (auto it = evse_profiles.begin(); it != evse_profiles.end(); it++) {
+            if (profile.id == it->id) {
+                evse_profiles.erase(it);
+                found_profile = true;
+                break;
+            }
+        }
 
-    auto& profile_storage = (STATION_WIDE_ID == evse_id) ? station_wide_charging_profiles : charging_profiles[evse_id];
-    auto profile_with_id =
-        std::find_if(profile_storage.begin(), profile_storage.end(),
-                     [&profile](const ChargingProfile existing_profile) { return profile.id == existing_profile.id; });
-
-    // K01.FR05 - replace non-ChargingStationExternalConstraints profiles if id exists.
-    if (profile_with_id != profile_storage.end() &&
-        profile_with_id->chargingProfilePurpose != ChargingProfilePurposeEnum::ChargingStationExternalConstraints) {
-        *profile_with_id = profile;
-    } else {
-        profile_storage.push_back(profile);
+        if (found_profile) {
+            break;
+        }
     }
 
+    charging_profiles[evse_id].push_back(profile);
+
     return response;
 }
 
-std::vector<ChargingProfile> SmartChargingHandler::get_profiles() {
-    auto all_profiles = station_wide_charging_profiles;
+std::vector<ChargingProfile> SmartChargingHandler::get_station_wide_profiles() const {
+    std::vector<ChargingProfile> station_wide_profiles;
+    try {
```

Comment:
I think we should be able to do this without needing an exception. Maybe check using `.count()` or `.find()` to check if the `STATION_WIDE_ID` is present.